core/bootstrap/src/main/java/org/wildfly/swarm/bootstrap/Main.java

```diff
@@ -34,16 +59,86 @@ public static void main(String... args) throws Throwable {
             Performance.start();
             //TODO Move property key to -spi
             System.setProperty(BootstrapProperties.IS_UBERJAR, Boolean.TRUE.toString());
+
+            String processFile = System.getProperty(MAIN_PROCESS_FILE);
+
+            if (processFile != null) {
+                shutdownService = Executors.newSingleThreadExecutor();
+                shutdownService.submit(() -> {
+                        File uuidFile = new File(processFile);
+                        try {
+                            register(uuidFile.getParentFile(), uuidFile.toPath());
+                            processEvents(uuidFile.toPath());
+                            if (mainInvoker != null) {
+                                mainInvoker.stop();
+                            }
+                            //Exit gracefully
+                            System.exit(0);
+                        } catch (Exception e) {
+                            e.printStackTrace();
+                        }
+
+                        return null;
+                    }
+                );
+            }
+
             new Main(args).run();
         } catch (Throwable t) {
             t.printStackTrace();
             throw t;
         }
     }
 
+    private static void register(File directory, Path file) throws IOException {
+        watcher = FileSystems.getDefault().newWatchService();
+        keys = new HashMap<WatchKey,Path>();
```

Comment:
This always replaces the `keys` map with a new one, so the code as is only works with a single `Key` by necessity. In that case, it could probably be drastically simplified.